hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/staging/Paths.java

```diff
@@ -0,0 +1,306 @@
+        uuid,
+        Integer.toString(getAppAttemptId(conf)),
+        attempt.toString());
+  }
+
+  /**
+   * Try to come up with a good temp directory for different filesystems.
+   * @param fs filesystem
+   * @param conf configuration
+   * @return a path under which temporary work can go.
+   */
+  public static Path tempDirForStaging(FileSystem fs,
+      Configuration conf) {
+    Path temp;
+    switch (fs.getScheme()) {
+    case "file":
+      temp = fs.makeQualified(
+          new Path(System.getProperty(JAVA_IO_TMPDIR)));
+      break;
+
+    case "s3a":
+      // the Staging committer may reject this if it doesn't believe S3A
+      // is consistent.
+      temp = fs.makeQualified(new Path(FILESYSTEM_TEMP_PATH));
+      break;
+
+    // here assume that /tmp is valid
+    case "hdfs":
+    default:
+      String pathname = conf.getTrimmed(
+          FS_S3A_COMMITTER_STAGING_TMP_PATH, FILESYSTEM_TEMP_PATH);
+      temp = fs.makeQualified(new Path(pathname));
+    }
+    return temp;
+  }
+
+  /**
+   * Get the Application Attempt ID for this job.
+   * @param conf the config to look in
+   * @return the Application Attempt ID for a given job.
+   */
+  private static int getAppAttemptId(Configuration conf) {
+    return conf.getInt(
+        MRJobConfig.APPLICATION_ATTEMPT_ID, 0);
+  }
+
+  /**
+   * Build a temporary path for the multipart upload commit information
+   * in the supplied filesystem (which is expected to be the cluster FS)
+   * @param conf configuration defining default FS.
+   * @param uuid uuid of job
+   * @return a path which can be used for temporary work
+   * @throws IOException on an IO failure.
+   */
+  public static Path getMultipartUploadCommitsDirectory(Configuration conf,
+      String uuid) throws IOException {
+    FileSystem fs = FileSystem.get(conf);
+    return getMultipartUploadCommitsDirectory(fs, conf, uuid);
+  }
+
+  /**
+   * Build a temporary path for the multipart upload commit information
+   * in the supplied filesystem (which is expected to be the cluster FS)
+   * @param fs target FS
+   * @param conf configuration
+   * @param uuid uuid of job
+   * @return a path which can be used for temporary work
+   * @throws IOException on an IO failure.
+   */
+  static Path getMultipartUploadCommitsDirectory(FileSystem fs,
+      Configuration conf, String uuid) throws IOException {
+    return path(tempDirForStaging(fs, conf),
+        UserGroupInformation.getCurrentUser().getShortUserName(),
+        uuid,
+        STAGING_UPLOADS);
+  }
+
+  // TODO: verify this is correct, it comes from dse-storage
+  private static Path localTemp(Configuration conf, int taskId, int attemptId)
+      throws IOException {
+    String[] dirs = conf.getTrimmedStrings(BUFFER_DIR);
+    Random rand = new Random(Objects.hashCode(taskId, attemptId));
+    String dir = dirs[rand.nextInt(dirs.length)];
+
+    return FileSystem.getLocal(conf).makeQualified(new Path(dir));
```

Comment:
We ended up changing this to balance across multiple local directories using `LocalDirAllocator`. That requires keeping track of the directory used for each task attempt in a static hashmap to ensure you don't lose track of files. Here's what we use now:

```java
  @VisibleForTesting
  public static Cache<TaskAttemptID, Path> tempFolders = CacheBuilder
      .newBuilder().build();

  public static Path getLocalTaskAttemptTempDir(final Configuration conf,
                                                final String uuid,
                                                final TaskAttemptID attemptID)
      throws IOException {
    try {
      return tempFolders.get(attemptID, new Callable<Path>() {
        @Override
        public Path call() throws Exception {
          LocalDirAllocator allocator = new LocalDirAllocator(
              "mapreduce.cluster.local.dir");
          return FileSystem.getLocal(conf).makeQualified(
              allocator.getLocalPathForWrite(uuid, conf));
        }
      });
    } catch (ExecutionException e) {
      throw new RuntimeException(e.getCause());
    } catch (UncheckedExecutionException e) {
      if (e.getCause() instanceof RuntimeException) {
        throw (RuntimeException) e.getCause();
      }
      throw new RuntimeException(e);
    }
  }
```

Reply:
OK, I'll do that. 
1. Any tests?
2. Task committer will invalidate the cache value for its task ID at end of taskCommit